api/v1alpha1/ngrok_common.go

```diff
@@ -162,3 +163,252 @@ type EndpointSAML struct {
+	// allowed access to the endpoint
+	EmailDomains []string `json:"emailDomains,omitempty"`
+}
+
+func (opc OAuthProviderCommon) toNgrokEndpointOauth() *ngrok.EndpointOAuth {
+	return &ngrok.EndpointOAuth{
+		OptionsPassthrough: opc.OptionsPassthrough,
+		CookiePrefix:       opc.CookiePrefix,
+		InactivityTimeout:  uint32(opc.InactivityTimeout.Duration.Seconds()),
+		MaximumDuration:    uint32(opc.MaximumDuration.Duration.Seconds()),
+		AuthCheckInterval:  uint32(opc.AuthCheckInterval.Duration.Seconds()),
+		Provider:           ngrok.EndpointOAuthProvider{},
+	}
+}
+
+func (opc OAuthProviderCommon) ClientSecretKeyRef() *SecretKeyRef {
+	return opc.ClientSecret
+}
+
+type EndpointOAuth struct {
+	// configuration for using github as the identity provider
+	Github *EndpointOAuthGitHub `json:"github,omitempty"`
+	// configuration for using facebook as the identity provider
+	Facebook *EndpointOAuthFacebook `json:"facebook,omitempty"`
+	// configuration for using microsoft as the identity provider
+	Microsoft *EndpointOAuthMicrosoft `json:"microsoft,omitempty"`
+	// configuration for using google as the identity provider
+	Google *EndpointOAuthGoogle `json:"google,omitempty"`
+	// configuration for using linkedin as the identity provider
+	Linkedin *EndpointOAuthLinkedIn `json:"linkedin,omitempty"`
+	// configuration for using gitlab as the identity provider
+	Gitlab *EndpointOAuthGitLab `json:"gitlab,omitempty"`
+	// configuration for using twitch as the identity provider
+	Twitch *EndpointOAuthTwitch `json:"twitch,omitempty"`
+	// configuration for using amazon as the identity provider
+	Amazon *EndpointOAuthAmazon `json:"amazon,omitempty"`
+}
+
+type EndpointOAuthGitHub struct {
+	OAuthProviderCommon `json:",inline"`
+	// a list of github teams identifiers. users will be allowed access to the endpoint
+	// if they are a member of any of these teams. identifiers should be in the 'slug'
+	// format qualified with the org name, e.g. org-name/team-name
+	Teams []string `json:"teams,omitempty"`
+	// a list of github org identifiers. users who are members of any of the listed
+	// organizations will be allowed access. identifiers should be the organization's
+	// 'slug'
+	Organizations []string `json:"organizations,omitempty"`
+}
+
+func (github *EndpointOAuthGitHub) ToNgrok(clientSecret *string) *ngrok.EndpointOAuth {
+	if github == nil {
+		return nil
+	}
+
+	mod := github.toNgrokEndpointOauth()
+	mod.Provider.Github = &ngrok.EndpointOAuthGitHub{
+		ClientID:       github.ClientID,
+		ClientSecret:   clientSecret,
+		Scopes:         github.Scopes,
+		EmailAddresses: github.EmailAddresses,
+		EmailDomains:   github.EmailDomains,
+		Teams:          github.Teams,
+		Organizations:  github.Organizations,
+	}
+	return mod
+}
+
+type EndpointOAuthFacebook struct {
+	OAuthProviderCommon `json:",inline"`
+}
+
+func (facebook *EndpointOAuthFacebook) ToNgrok(clientSecret *string) *ngrok.EndpointOAuth {
+	if facebook == nil {
+		return nil
+	}
```

Comment:
are these checks necessary? I guess i'm just not used to adding them, but is it just a good safety check?

Reply:
Since every type of provider is optional and can be `nil`, when we try to call `ToNgrok(*string)` if they are nil, we'll get a panic, so I think this is a good defensive check. We also check to see if its nil in the edge controller and if it is, we skip the provider.